File: ChatGPT/Intermediárias/sabbachard/Sabbachard.java
public class Sabacchard {
    public static int maxScore(int[] cards) {
        int n = cards.length;
        int[][] dp = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            dp[i][i] = cards[i];
        }
        
        for (int length = 2; length <= n; length++) {
            for (int i = 0; i <= n - length; i++) {
                int j = i + length - 1;
                dp[i][j] = Math.max(cards[i] + Math.min((i + 2 <= j ? dp[i + 2][j] : 0),
                                                       (i + 1 <= j - 1 ? dp[i + 1][j - 1] : 0)),
                                   cards[j] + Math.min((i + 1 <= j - 1 ? dp[i + 1][j - 1] : 0),
                                                       (i <= j - 2 ? dp[i][j - 2] : 0)));
            }
        }
        
        return dp[0][n - 1];
    }

    public static void main(String[] args) {
        int[] cards = {10, 19, 1, 2};
        System.out.println(maxScore(cards));  // Output: 21
    }
}

File: ChatGPT/Intermediárias/sabbachard/Sabbachard.cpp
#include <vector>
#include <algorithm>
#include <iostream>

int maxScore(const std::vector<int>& cards) {
    int n = cards.size();
    std::vector<std::vector<int>> dp(n, std::vector<int>(n, 0));
    
    for (int i = 0; i < n; ++i) {
        dp[i][i] = cards[i];
    }
    
    for (int length = 2; length <= n; ++length) {
        for (int i = 0; i <= n - length; ++i) {
            int j = i + length - 1;
            dp[i][j] = std::max(cards[i] + std::min(dp[i + 2][j], dp[i + 1][j - 1]),
                               cards[j] + std::min(dp[i + 1][j - 1], dp[i][j - 2]));
        }
    }
    
    return dp[0][n - 1];
}

int main() {
    std::vector<int> cards = {10, 19, 1, 2};
    std::cout << maxScore(cards) << std::endl;  // Output: 21
    return 0;
}
